File: leetcode/30-day-leetcoding-challenge/30-day-leetcoding-challenge/week-2/day-10-min-stack/stack-of-pairs.py
class MinStack:

    def __init__(self):
        """
        initialize your data structure here.
        """
        self.minStack = []
        

    def push(self, x: int) -> None :
        if not self.minStack : self.minStack.append((x , x))
        else : 
            self.minStack.append((x , min(x , self.minStack[-1][1])))
        

    def pop(self) -> None:
        if self.minStack :
            self.minStack.pop()
        

    def top(self) -> int :
        return self.minStack[-1][0]
        

    def getMin(self) -> int :
        return self.minStack[-1][1]
        


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(x)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()

File: leetcode/30-day-leetcoding-challenge/30-day-leetcoding-challenge/week-2/day-10-min-stack/stack-of-pairs.cpp
class MinStack {
public:
    /** initialize your data structure here. */
    stack<int>general ;
    stack<pair<int , int>>minStack ;
    MinStack() 
    {
        
    }
    
    void push(int x) 
    {
        if(minStack.empty())
        {
            minStack.push({x , x}) ;
        }
        else
        {
            int minEle = (minStack.top()).second ;
            minStack.push({x , min(x , minEle)}) ;
        } 
    }
    
    void pop() 
    {
        if(!minStack.empty())
            minStack.pop() ;
    }
    
    int top() 
    {
        return (minStack.top()).first ;
    }
    
    int getMin() 
    {
        return (minStack.top()).second ;
    }
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(x);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */
